rp2040-hal/examples/dormant_sleep.rs

```diff
@@ -0,0 +1,295 @@
+//! # DORMANT low-power mode example
+//!
+//! This application demonstrates how to enter and exit the RP2040's lowest-power DORMANT mode
+//! where all clocks and PLLs are stopped.
+//!
+//! Pulling GPIO 14 low (e.g. via a debounced momentary-contact button) alternately wakes the
+//! RP2040 from DORMANT mode and a regular WFI sleep. A LED attached to GPIO 25 (the onboard LED
+//! on the Raspberry Pi Pico) pulses once before entering DORMANT mode and twice before entering WFI sleep.
+//!
+//! It may need to be adapted to your particular board layout and/or pin assignment.
+//!
+//! See the `Cargo.toml` file for Copyright and license details.
+
+#![no_std]
+#![no_main]
+
+#[allow(unused_imports)]
+use panic_halt as _;
+
+use rp2040_hal as hal;
+
+use core::{cell::RefCell, ops::DerefMut};
+
+use critical_section::Mutex;
+
+use embedded_hal::digital::v2::ToggleableOutputPin;
+
+use fugit::RateExtU32;
+
+use hal::{
+    clocks::{ClockError, ClocksManager, InitError, StoppableClock},
+    gpio,
+    gpio::{Interrupt::EdgeLow, Pins},
+    pac,
+    pac::{interrupt, CLOCKS, PLL_SYS, PLL_USB, RESETS, XOSC},
+    pll::{
+        common_configs::{PLL_SYS_125MHZ, PLL_USB_48MHZ},
+        setup_pll_blocking, start_pll_blocking, Disabled, Locked, PhaseLockedLoop,
+    },
+    rosc::RingOscillator,
+    sio::Sio,
+    watchdog::Watchdog,
+    xosc::{setup_xosc_blocking, CrystalOscillator, Stable, Unstable},
+    Clock,
+};
+
+use nb::block;
+
+type ClocksAndPlls = (
+    ClocksManager,
+    CrystalOscillator<Stable>,
+    PhaseLockedLoop<Locked, PLL_SYS>,
+    PhaseLockedLoop<Locked, PLL_USB>,
+);
+
+type RestartedClockAndPlls = (
+    CrystalOscillator<Stable>,
+    PhaseLockedLoop<Locked, PLL_SYS>,
+    PhaseLockedLoop<Locked, PLL_USB>,
+);
+
+/// The button input.
+type ButtonPin = gpio::Pin<gpio::bank0::Gpio14, gpio::FunctionSioInput, gpio::PullNone>;
```

Comment:
I needed to activate a pull-up on gpio14 to make the example work:
```suggestion
type ButtonPin = gpio::Pin<gpio::bank0::Gpio14, gpio::FunctionSioInput, gpio::PullUp>;
```

With that, it works as expected, and I see the current drawn by the pico to alternate between 0.03A and 0.05A (using a cheap USB power monitor with 10mA resolution). Nice!